### DISCUSSION TITLE
[API Request] Expose codec, container, full codec

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#87] [API requests]
### Type of change

API extension

### Proposal description

Now that we split the query in 2 and already have bitrates and samplerate let's finish the missing data.

Expose new fields for songs:
- codec: Short well known codec name (pcm, mp3, ...)
- container: The container (webm, mkv, dsd, ...)
- full codec: The full codec name with details like the different variations of pcm

### Backward compatibility

- [X] No backward compatibility impact.

### API details

TBD depending on the choice made.

### Potential issues

As discussed before this adds more data to the json payloads, but this greatly help apps to decide to request transcoding or not.

The simple solution would be to only return the extended data in the `getSong` endpoint but for the moment we have no process / documentation solution to indicate that a field for a response should only be filled for a specific endpoint so it requires to define that too if we go that road.

**Comment by dweymouth** [comment]
I think adding *full* codec info to the base payloads might be not desirable depending on how much info we want to add there, and introducing a new endpoint `getCodecInfo` that takes multiple song IDs and returns the long-form codec info for each could be a way to go. The short-form codec and container could be made part of the base song payload.

**Reply by Tolriq** [reply]
Well IMO the simple solution is to do as the other media providers and just expose the codec and container values as exposed by ffmpeg.
So that clients have everything to decide if they should transcode or not and can display useful info to users.

**Reply by deluan** [reply]
What exactly do you mean by "full codec info"? Something like this (from `ffprobe`):
```
        {
            "index": 0,
            "codec_name": "flac",
            "codec_long_name": "FLAC (Free Lossless Audio Codec)",
            "codec_type": "audio",
            "codec_tag_string": "[0][0][0][0]",
            "codec_tag": "0x0000",
            "sample_fmt": "s16",
            "sample_rate": "44100",
            "channels": 2,
            "channel_layout": "stereo",
            "bits_per_sample": 0,
            "initial_padding": 0,
            "r_frame_rate": "0/0",
            "avg_frame_rate": "0/0",
            "time_base": "1/44100",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 10932096,
            "duration": "247.893333",
            "bits_per_raw_sample": "16",
            "extradata_size": 34,
            "disposition": {
                "default": 0,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0,
                "non_diegetic": 0,
                "captions": 0,
                "descriptions": 0,
                "metadata": 0,
                "dependent": 0,
                "still_image": 0
            }
        },
```

If that's the case, I totally agree with @dweymouth that this is too much to add in the child element.

**Reply by Tolriq** [reply]
No it was based on the request of Dweymouth to have a simpler codec, so it would require a full one.
Like all the variations of pcm.
If we return the ffmpeg codec untouched then there's no need for more data.

**Reply by deluan** [reply]
So can you give a example of what do you expect as "full codec info" as reported by `ffmpeg`, and add it to the description?

**Reply by epoupon** [reply]
We can get the complete list with `ffmpeg -decoders` and `ffmpeg -demuxers`

Here is the list of decoders I have encountered with my test files.

Decoders:
* aac
* alac
* ape
* dsd_lsbf_planar (there are also dsd_lsbf, dsd_msbf, dsd_msbf_planar)
* flac
* pcm_s16_le (there are a lot of variants, can't list them all here)
* musepack7 (there is also musepack8)
* opus
* vorbis
* mp3
* wavpack
* wmav2 (there are also wmav1, wmalossless, wmapro, and wmavoice)

As for demuxers:
* aac
* ape
* asf
* aiff
* dsf
* matroska,webm (yes, this is really the name of this single demuxer)
* mov,mp4,m4a,3gp,3g2,mj2 (yes, this is really the name of this single demuxer)
* mp3
* mpc
* ogg
* wv

Some decoders are problematic because of their variants, some demuxers are problematic because they are used to handle several file formats.
I think there are not that many decoders/demuxers: we could probably handle all use cases by defining file formats/codecs in our API, and have a 'variant' field to set the extra variant info if needed.
What do you think?

**Reply by Tolriq** [reply]
FFmpeg adds new codecs and demuxers on each of their release. We can't build an API on fixed values. (Your list misses ac3, eac3, dts, iso container and quite a few other common ones).

You need to see the transcoding API and needs in it's globality and the purpose of those values.

A transcoding API as it works in other media providers and it well proven as working would be something like:
The client send a request to the transcoding decision engine with a profile and the server answers with what it can do.

The profile would contain:
I can direct play the codecs x,y,z,w in container a,b,c with limitation for codec X or container A to max channel count 2 and max sample rate 96000.
I want transcoded media in codec X containers A with limitations .... or .... or .... (in order of preferences).

The server answers:
- You can direct play go for it.
- I can transcode the selected profile is X with the params the clients needs to pass to get that profile
- You can't direct play and I can't transcode.

Since the supported codecs on the client side needs the actual real codec and not the generic one it's important to be able to discuss with the proper detailed codec values. Not all clients supports all pcm variations, same for wma, and same for any future codec that may arrise.

There's even advanced codec details that can arise and should be taken in account like aac xHE that is not supported by default by ffmpeg (but can with some external lib) and yet is supported by Android, so the server could make a mistake to assume it can't handle the file when the client can direct play.

So TL;DR this is a complex need, and it requires deep codec details and media details to properly handle.

**Reply by epoupon** [reply]
Sure, but I don't understand why defining our own list of formats/codecs would prevent us from doing what you propose?
Maybe I missed something, but do we really want to blindly pass demuxers and decoders to the clients? What about "combined " demuxers? What if tomorrow somebody adds an experimental codec with some custom name?

**Reply by Tolriq** [reply]
If we define a list we need to handle all the codec details anyway so we will have a 1 to 1 mapping of names only subsonic use, this brings 0 benefits and only troubles.
We need to maintain the list and version it so clients and servers can communicate on those arbitrary values.
So now the clients needs to query what version of the codec list the server support, then internally have a list of the codecs then do a mapping of the codecs he knows he support to the list of the codecs the server support that is not common to anything he can gather via ffmpeg if he use that to play as most do.
And in all case servers will use ffmpeg to transcode so will also need to have the mapping in the other side to do the transcoding..

So the real question, if why do you want to use something specific that brings absolutely no benefits for the server or the clients?

**Reply by deluan** [reply]
IMHO, we should use the same pattern already well-stablished by the most used protocol on earth: HTTP.

When calling `/stream`, clients can pass one or multiple "accept" headers with format/codec/containers it can handle, in order of preference. Then servers can select the proper transcoding to use from that list.

Anyway, I agree that we should not pass a full list to the client for it to decide what to do. This complicates client implementations with implementation details that are better handled by the server.

**Reply by epoupon** [reply]
> So the real question, if why do you want to use something specific that brings absolutely no benefits for the server or the clients?

Because the topic here is to "expose codec, container, full codec". Clients will get containers  like "mov,mp4,m4a,3gp,3g2,mj2", this is just wtf.
But if this is just to define a ffmpeg-based transcoding engine, fine!

**Reply by deluan** [reply]
I think this discussion is confusing as it is trying to define two different needs (as @Tolriq said [here](https://github.com/opensubsonic/open-subsonic-api/discussions/87#discussioncomment-9360167)):
- Parameters for the transcoding engine
- Expose format/codec/container to the clients, so they can show this info.

My suggestion for the "accept" header is for the first one. For the second one (expose), I think that's hard as @epoupon pointed out, given that most users use `TagLib` for scanning.

**Reply by Tolriq** [reply]
@deluan again no, because client A might request a streaming command for a device B that does not support headers, like UPnP devices or Chromecast.
This is the same issue as auth for streaming.

And again there's different needs between direct play and transcoding and complex constraints to take in account.

A client can play different codecs and containers with different constraints that's already something that is nearly impossible to express with a simple accept header.

And a client want to control what it want for transcoding with order of preferences, to priorize a format for higher quality for a given BW or because the device have a native decoder that will use less battery or because the format is better for streaming in low network quality.

So the API needs to be more or less as described if we want a proper API with server side transcoding engine. send a list of supported direct play, send a list of prefered transcoding profile and server answers with the proper choice.

The other solution is to delegate all on client side, an endpoint to get the exact full codec container and everything from ffprobe, then the client make the decision and send transcoding command with all the params to the server.

The problem of doing it client side is that clients don't know how the ffmpeg of the server is compiled, so can't know what codecs and containers are supported and what specific option it should pass. So it will still require to define a communication protocol to pass the data and have the server convert to ffmpeg data.

Hence why the server side transcoding decision engine makes more sense.

Emby, Plex, Jellyfin all works more or less like that and support endless combination of clients on endless combination of devices and casting and everything and it fit all the needs of all the cases since ages.

Here's a old example of a Plex client profile (it's XML but it can be send via the API too)

It also contains video and it's used server side to auto select the profile based on the UPnP device that connect to the server hence the identification part.

But this is the kind of things we need to be able to support to finally be on part with the rest of the media servers and have a proper complete solution for streaming and casting and supporting all clients in the best possible ways.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Client name="PlayStation 4">
   <!-- Author: Plex Inc. -->
   <Identification>
	  <Header name="User-Agent" substring="UPnP/1.0 DLNADOC/1.50" />
	  <Header name="User-Agent" substring="PS4Application" />
	  <Header name="User-Agent" substring="PlayStation 4" />
	  <Header name="X-AV-Client-Info" substring="PlayStation 4" />
   </Identification>
   <TranscodeTargets>
   	  <VideoProfile container="mpegts" audioCodec="aac,ac3" codec="h264">
		 <Setting name="VideoEncodeFlags" value="-x264opts cabac=1" />
	  </VideoProfile>
	  <MusicProfile container="mp3" codec="mp3" />
	  <PhotoProfile container="jpeg" />
   </TranscodeTargets>
   <DirectPlayProfiles>
	  <VideoProfile container="mpegts" codec="mpeg2video,h264,mpeg4" audioCodec="ac3,aac,mp3" />
	  <VideoProfile container="mp4" codec="mpeg2video,h264,mpeg4" audioCodec="ac3,aac,mp3" />
	  <VideoProfile container="mov" codec="mpeg2video,h264,mpeg4" audioCodec="ac3,aac,mp3" />
	  <VideoProfile container="mkv" codec="mpeg2video,h264,mpeg4" audioCodec="ac3,aac,mp3" />
	  <VideoProfile container="avi" codec="mpeg2video,h264,mpeg4" audioCodec="ac3,aac,mp3" />
	  <MusicProfile container="mp3" codec="mp3" />
	  <MusicProfile container="mp4" codec="aac" />
	  <PhotoProfile container="jpeg,bmp,png" />
   </DirectPlayProfiles>
   <CodecProfiles>
	  <VideoCodec name="mpeg2video">
		 <Limitations>
			<UpperBound name="video.width" value="1920" />
			<UpperBound name="video.height" value="1080" />
			<UpperBound name="video.bitrate" value="20000" />
		 </Limitations>
	  </VideoCodec>
	  <VideoCodec name="h264,mpeg4">
		 <Limitations>
			<UpperBound name="video.width" value="1920" />
			<UpperBound name="video.height" value="1080" />
			<UpperBound name="video.bitrate" value="20000" />
		 </Limitations>
	  </VideoCodec>
	  <VideoAudioCodec name="ac3,aac">
		 <Limitations>
			<UpperBound name="audio.channels" value="6" />
		 </Limitations>
	  </VideoAudioCodec>
	  <VideoAudioCodec name="mp3">
		 <Limitations>
			<UpperBound name="audio.channels" value="2" />
		 </Limitations>
	  </VideoAudioCodec>
	  <MusicCodec name="aac">
		 <Limitations>
			<UpperBound name="audio.channels" value="2" />
		 </Limitations>
	  </MusicCodec>
   </CodecProfiles>
   <ContainerProfiles>
	  <PhotoContainer name="*">
		 <Limitations>
			<UpperBound name="media.width" value="1920" />
			<UpperBound name="media.height" value="1080" />
		 </Limitations>
	  </PhotoContainer>
   </ContainerProfiles>
</Client>
```

As a simple personal need, I often cast from a Subsonic server to my Sonos devices, the sonos support flac in high quality but up to some bounds on sample rate / bits per sample.

So I need direct play up to that bound then transcoding to flac to the upper bound to reduce the quality impact all that controlled by Symfonium. 

Edit: For last deluan message, they are 2 different needs but they are related in the end as it impact if there will be transcoding and it's part of the reason why the user do want to know the data before playing.

**Reply by epoupon** [reply]
> send a list of supported direct play

What do you mean by that? A list of decoders/demuxers? Is it from its local ffmpeg?

**Reply by Tolriq** [reply]
It's shown as example in the profile from plex in the post and explained earlier. The client knows what it can play so it can build a profile, it's up to the client to know the name of the formats and containers it support.

So again as example with only music:

```
   <DirectPlayProfiles>
	  <MusicProfile container="mp3" codec="mp3" />
	  <MusicProfile container="mp4" codec="aac" />
   </DirectPlayProfiles>
   <CodecProfiles>
	  <VideoAudioCodec name="mp3">
		 <Limitations>
			<UpperBound name="audio.channels" value="2" />
		 </Limitations>
	  </VideoAudioCodec>
	  <MusicCodec name="aac">
		 <Limitations>
			<UpperBound name="audio.channels" value="2" />
		 </Limitations>
	  </MusicCodec>
   </CodecProfiles>
```

That client support mp3 in mp3 container and aac in mp4 container. And for the codecs a max channel of 2.
The client would also pass a global maxBW in the profile.

With that the server knows if the file can be directly played or not.

If the profile also contains transcoding profile then the server can answer that it can transcode and return the appropriate data to pass to the stream endpoint to reach what the client desire.

**Reply by epoupon** [reply]
Ok! Now how does the client know what it can play? What are the expected values for the containers and codecs the client send to the server?

**Reply by deluan** [reply]
I'm confused where/how this "profile" will be used? Will this be part of the request? What exactly are you proposing the server would receive in the `/stream` endpoint?

I still think this logic should be on the server, and the client just specifies what it supports. And the `accept` "header" can be passed as query params as well, to cater for the device casting.

**Reply by Tolriq** [reply]
Can you please read all I write and not just extracts?

1) The client obviously knows what it can play since it's itself. I know what I can play. The containers and codecs are the well known values from ffmpeg if that's what we decide. If you want to build something more complex and harder on all side then why not.

2) The profile is sent to the transcoding decision endpoint, that answer with the necessary information for the client to request the proper streaming url be it direct play or transcoded, it's already written twice in my long explanations.

> And the accept "header" can be passed as query params as well,

Please show me how you pass all the necessary information? A list of supported containers and codecs with limitations on each of them + an ordered list of wanted transcoded profile with again container and codecs and limitations on them.

**Reply by deluan** [reply]
> it's already written twice in my long explanations.

TL;DR 😛

> transcoding decision endpoint

So this would be a call that returns a stream URL? That is starting to make sense 🙂

> Please show me how you pass all the necessary information?

```
/stream?accept=<JSON object with the profile>
```

But if we are talking about having a new endpoint to get the stream URL, then this is not necessary.

**Reply by epoupon** [reply]
Ok clients send what they can play using their local ffmpeg demuxers/decoders information.
In your example, you advertise mp4 support, but depending of your ffmpeg version, it may actually use the "mov,mp4,m4a,3gp,3g2,mj2" demuxer. So clients and servers needs to further parse/split demuxers considering it is comma separated?

**Reply by Tolriq** [reply]
The call would return something to be defined :)

Imo directly returning the streaming url is limiting to help the client display the proper details to the client.

I would return an object with a result that describe if it can direct play (and eventually the direct play url to simplify), or it can transcode with an url to stream + the details about what format was selected to transcode so the client can display it.
Or an error if it can't direct play and can't transcode due to broken profile or ffmpeg not supporting the request.

And since we return to the display the format it's transcoded too we have circled again to the name of the codecs and containers :) 
Hence why I think they are linked and using the ffmpeg values makes sense as a whole for now.

Edit: And I remind you about limits on header and url size you do not pass large json like that.

**Reply by Tolriq** [reply]
Client not always use ffmpeg, but they do know what they support. 
Ffmpeg understand mp4 by itself so there's no need for that one to do anything. 

But in all cases yes the server will need to do some sanitization of the profile to be able to return a proper error and not an invalid streaming url.

**Reply by epoupon** [reply]
By the way, with the server centric transcoding approach, do we really *need* to expose all these information for each song?
Clients just issue a stream request with their profile as said before, and the server can analyze the files only during the request to know what to do?

**Reply by Tolriq** [reply]
It's information that users like to have for display purpose and for filtering (as said too :p)

Typically they can make smart playlists for HQ audio, filter low quality media and many other things.

**Reply by epoupon** [reply]
Ah yes indeed, missed that one. Now you know we really don't read you :)

And maybe I missed this too, but what is "full codec info" vs codec+container?

**Reply by Tolriq** [reply]
Well it was said too :)

It was the first version of the proposal that gave a limited know list of codecs, and so needed a way to provide the complete one.

**Reply by epoupon** [reply]
Ok so now? we remove this info and just focus on container/codec?

**Reply by Tolriq** [reply]
https://github.com/opensubsonic/open-subsonic-api/discussions/87#discussioncomment-9072369 ? :)

**Comment by Tolriq** [comment]
So to move on this I suggest we simplify and return to the very initial stuff of returning the codec and container as returned by ffmpeg.
They are well known and every server that will transcode will use ffmpeg so will use those values for the transcoding part.

Having consistent values in the returned data and for the future transcoding api makes things easy and simple.

ffprobe -formats for the containers
ffprobe -codecs for the codecs

We do not enforce any list as new codecs can be added in future ffmpeg versions.

**Reply by epoupon** [reply]
I am bit uncomfortable to rely on ffmpeg for this, as a server may not use it.
Furthermore LMS only uses taglib to scan files, this would require a double parse to get the right codec info (or a hardcoded codec name map to avoid that, if tag lib provides enough information). Will do some tests...

**Reply by Tolriq** [reply]
Tag lib does not return any actual proper value and is ultra limited in what in support.
Precise data is needed for transcoding and transcoding will require ffmpeg anyway.

Edit: For raw limited display only purpose we already have a field in the API.

**Reply by deluan** [reply]
I agree with @epoupon that forcing the server to use `ffmpeg` to scan files is not desirable. I think this could be implemented as a new endpoint that receives a list of ids and return the codec info for those ids only. This can be cached in the server for future requests. But having to process this beforehand (during import) can make the scan process really slow for servers that rely on `taglib` (most of them?)

**Reply by Tolriq** [reply]
Taglib does not return proper data to be used.

We can rely on a new endpoint but for one song, if we allow multiple then I ask for the 300k in one shot and it's not working.

And at that point we can directly build the transcoding API that goes with it.

**Reply by deluan** [reply]
> Taglib does not return proper data to be used.

Yes, I understand that, but don't think forcing the server to run ffprobe/ffmpeg on every scanned song is a solution here.

> We can rely on a new endpoint but for one song, if we allow multiple then I ask for the 300k in one shot and it's not working.

I think the idea with for this endpoint (suggested by @dweymouth) would be to call it when you add songs to the playqueue, so the client can add the proper transcoding parameters at that moment.

> And at that point we can directly build the transcoding API that goes with it.

I personally think this should be the way to go, instead of debating on a half-backed solution.

**Reply by Tolriq** [reply]
You do not need to prepare 300 songs in advance you need to transcode the one you play. Preparing all the queue just slow down playback on the client. It's the same for the lyrics you don't need to extract all the queue lyrics on the start.

And there's 2 different needs, the transcoding engine, but also exposing actual data that users can use for filtering.
All other non subsonic providers expose proper codec and container for that purpose.
We now have sample rate and others but still miss proper codecs for that need.

So if no ffmpeg then that need is not possible to fit as we can't do 300000 request to get all the data one by one.

Anyway the real urgency is the transcoding part as it's really limiting Subsonic but exposing proper codec is a frequent user need that should be tacked at some point.
